만국박람회 STEP2 Kyo, 토털이

tottalE opened the pull request:
@lina0322
안녕하세요 엘림! STEP2 완료하고 PR드립니다!
가감없는 피드백 부탁드려요 🔥

### 진행 단계 : Step2
## ⭐️ 구현한 기능
- `DataManager` 
    - Decoder 역할을 하고, Data들을 ViewController에서 사용할 수 있도록 해주는 구조체입니다.
    - Generic으로 구성하였으며 단일 Data, Data List 들을 가지고 올수 있도록 2가지의 메서드를 구현하였습니다.
- `ExpoViewController`
    - App을 가장 처음 실행하면 보이는 화면을 담당하는 ViewController입니다.
- `KoreaItemsViewController`
    - Item의 List를 보여주는 ViewController입니다. 
    - 해당 부분의 TableView Cell을 보여줄 때, CustomCell이 아닌 기본 Cell을 이용하여 구현하였습니다. 직접적인 제약조건은 줄 수 없었지만 이미지의 크기, reservedLayoutSize(이미지가 넉넉하게 차지할 공간), Text와 Image간의 Padding을 이용하여 제약조건을 간접적으로 줄 수 있었습니다.
- `DetailViewController`
    - 마지막에 Item을 클릭했을 때의 상세화면입니다.
    - 해당 화면은 코드를 사용한 경험을 하고 싶어 코드로 작성을 해보았습니다.

## ⭐️ 코드를 작성할 때 고민되었던 점
### 1. defaultContentConfiguration의 reservedLayoutSize
- 이번 프로젝트에서 custom cell이 아닌 기본셀을 공부해 보고 싶어 기본셀을 사용해 주었습니다. 그 과정에서 기존의 텍스트 요소에 접근한뒤 text를 바꾸어주는 방식은 depreciated 되어 defaultContentConfiguration를 사용하여 셀의 요소를 조정해주는 방법을 사용해 보았습니다. 새롭게 바뀐 부분이라 참고자료가 부족하였고, 많은 삽질 끝에 완성할 수 있었습니다.
- 이미지 사이즈를 조정해 주기 위한 요소를 찾던중, imageProperties의 reservedLayoutSize를 알게 되었습니다. 이미지를 위한 레이아웃을 예약 해 둔후 이미지를 가운데 위치시켜주는 기능인데, maximumSize 설정을 같은 크기로 해주어 이미지를 위한 공간을 만든 후 가운데 위치시켜 주었습니다. 이 과정에서 완벽히 이해하지 못한 부분이 있다면 피드백 해주시면 감사하겠습니다.

### 2. ScrollView의 contentLayoutGuide와 frameLayoutGuide
- contentLayoutGuide와 frameLayoutGuide, 그냥 scrollView와의 명확한 관계가 헷갈렸습니다. 구현은 모두 되었지만 그 이후에 contentLayoutGuide와 FrameLayout간의 관계, scrollView와 scrollView.ContentLayout 간의 차이점이 고민되었습니다.

## ⭐️ 해결이 되지 않은 점
- 첫번째 화면의 제목에서 (를 기준으로 개행을 주고 싶었는데 아직 좋은 방법을 찾지 못했습니다..!
  Step-3에서 조금 더 고민해보겠습니다.

## ⭐️ 조언을 얻고 싶은 부분

### 1. ScrollView의 Constraint
```swift
stackView.topAnchor.constraint(equalTo: scrollView.contentLayoutGuide.topAnchor),
stackView.bottomAnchor.constraint(equalTo: scrollView.contentLayoutGuide.bottomAnchor),
stackView.leadingAnchor.constraint(equalTo: scrollView.frameLayoutGuide.leadingAnchor),
stackView.trailingAnchor.constraint(equalTo: scrollView.frameLayoutGuide.trailingAnchor),
stackView.widthAnchor.constraint(equalTo: scrollView.widthAnchor),
```
- 위의 코드에서 `scrollView.contentLayoutGuide` 부분을 `scrollView`로 바꿔도 동일하게 동작이 되는 것을 확인했습니다.
- 제가 생각했을 때는 `scrollView`의 크기보다 늘어날수 있는 부분을 `scrollView.contentLayoutGuide`로 알고 있었는데 동작이 되서 조금 혼란스럽습니다. 둘의 차이점에 대해서 조언을 얻고 싶습니다.
➡️ 토털이가 찾아본 결과! `scrollView`를 superview(view)에 제약을 걸 때는  `scrollView.FrameLayoutGuide`로 제약이 걸리는 처럼 동작하고,
만약 `scrollView`와 `scrollView` 내부의 subView에 제약을 걸 때는 `scrollView.contentLayoutViewGuide`로 제약이 걸리는 것처럼 동작하게 됩니다!!.
그래서 명시적으로 `contentLayoutGuide`, `FrameLayoutGuide`를 표현해주는게 좋다고 생각이 들었습니다.

### 2. 코드로 UI를 짤 때의 구조
- 코드로 UI를 구성해준 후, `setUI()`와 `setConstraint()`로 메서드를 만들어주고, 익스텐션을 나누어 주었습니다. 코드로 UI를 구현해본 것이 처음이라 지금처럼 구성해주는 것이 맞는지 조언을 받고 싶습니다! 주로 사용하는 컨벤션이 있다던지, 익스텐션으로 나누어주는 기준이 있다던지 그런 것들이 궁금합니다!
- 또한 코드로 UI를 구성할 때 Constraint의 값을 관리할 때도 Namespace처럼 관리를 하는지 궁금합니다!

tottalE committed:
chore: view Controller 이름 ExpoViewController로 변경

tottalE committed:
chore: ViewController 이름 ExpoViewController로 변경

KyoPak committed:
feat : Main화면 Autolayout Constraint 적용

-  ScrollView 적용
-  각 Label, image, Button 추가

tottalE committed:
feat: Navigation Bar hidden ViewWillAppear에 적용

KyoPak committed:
feat : DataManager 구현 및 Error, AssetName namespace 생성

- DataManager 내부 Decode 역할 fetchExpoIntroData() 구현
- Error Type 별도로 DataError 생성
- File Name 별도로 AssetName 생성

tottalE committed:
feat: DataManager 내부 fetchExpoIntroData 함수내 로직 변경, throw로 에러처리를 할 수 있도록 함

KyoPak committed:
feat : ExpoViewController 내부 expoData, dataManager 프로퍼티 추가

tottalE committed:
feat: setData() 메서드 ExpoViewController 내에 생성

KyoPak committed:
feat : 메인화면 IBOutlet 연결 및 setupElementWithData() 구현

- 메인화면 IBOutlet 연결 및 Data setup 메서드 구현
- Constant namespace 생성
- AssetName에 타입 프로퍼티 추가

tottalE committed:
feat: ExpositionIntroduction 모델 내에 visitorsDescription, locationDescription, durationDescription 추가

- label에 들어갈 String을 computed property로 생성하여 바로 model struct로 부터 text label에 넣을 수 있도록 수정

KyoPak committed:
feat : ItemListViewController 생성 및 오토레이아웃 추가, 수정

- 폴더화
- ItemListViewController 생성
- 오토레이아웃 TableView, Cell 추가
- 오토레이아웃 Expo 관련 내부 텍스트 크기 변경

tottalE committed:
feat: KoreaItem JSON decoding 로직 추가, JSON Asset 이름 수정

KyoPak committed:
feat : nexButton 클릭 시 화면 넘어가는 기능 구현

- ItemListViewController ➡️ KoreaItemsViewController 이름 변경
- nextButtonTapped() IBAction 메서드 구현
- Identifer namespace 추가

tottalE committed:
feat: navigationBar hidden false로 변경, dataSource 및 delegate 정의, datasource내 tableView 메서드로 셀 내 데이터 삽입 및 셀 호출

KyoPak committed:
feat : KoreaItemList Cell UI 변경 및 네비게이션 바 타이틀, 바 아이템 color 변경

- Constant namespace 내부 프로퍼티 추가
- KoreaItemList Cell UI 내부 Font, Padding, Size 변경

tottalE committed:
feat: ExpoViewController내의 nextButton 에러 수정

KyoPak committed:
refactor : ExpoUI 내부 descriptionLabel ➡️ descriptionTextView 변경 및 AutoLayout 조정

- ExpoUI 내부 descriptionLabel ➡️ descriptionTextView로 변경
- descriptionTextView AutoLayout 조정

tottalE committed:
feat: ExpositionIntroduction 모델 내에 formatVisitorNumber() 메서드 추가, visitorsDescription에 적용, 에러 처리

KyoPak committed:
feat : DetailViewController로 화면 이동 구현 및 코드로 AutoLayout 구현

- DetailViewController 생성 및 코드로 AutoLayout 구현

tottalE committed:
feat: DetailView와 KoreaItemsView 사이의 데이터 전달 구현

KyoPak committed:
feat : Detail 화면 setData() 메서드 구현 및 AutoLayout 수정 구현

tottalE committed:
feat: navigation bar line 설정 및 title 설정

KyoPak committed:
refactor : DataManager 내부 Generic으로 수정 구현

- Decode 역할 하는 DataManager 내부 메서드 명 변경 및 Generic으로 수정 구현

tottalE committed:
refactor: ExpositionIntroduction 내부 visitor 프로퍼티명 visitorNumber로 수정, 메서드에도 적용

tottalE committed:
feat: Decodable으로 프로토콜 변경, visitors 네이밍 변경 및 CodingKey 추가

KyoPak committed:
refactor : DetailViewController 내부 프로퍼티 명 contentView ➡️ stackView 변경

tottalE committed:
feat: 접근제어자 추가

KyoPak committed:
refactor : DetailViewController 내부 setupContentView() ➡️ setupStackView() 메서드 명 변경 및 extension으로 메서드 분리

- setupContentView() ➡️ setupStackView() 메서드 명 변경
- extension으로 UI관련 메서드 분리

tottalE committed:
chore: textView bottom의  AutoLayout 제약 조건 추가 및 접근제어자 수정

tottalE committed:
Update README.md

tottalE committed:
Update README.md

tottalE committed:
Update README.md

tottalE committed:
Update README.md

KyoPak committed:
docx : ReadME 파일트리 추가

lina0322 commented on Expo1900/Expo1900/Base.lproj/Main.storyboard:
```diff
@@ -1,24 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="13122.16" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES" useSafeAreas="YES" colorMatched="YES" initialViewController="BYZ-38-t0r">
```
스토리보드를 사용하지 않으신 이유가 궁금하네요.👀

tottalE replied:
첫번째 화면, 두번째 화면은 스토리 보드를 사용해 주었고, 세번째 화면은 코드로 구현해 보고 싶은 욕심이 있어 코드로 구현해 주었습니다!

lina0322 replied:
단순히 코드로 구현해보고 싶은것이었나용?😀
코드와 스토리보드 각각의 장단점은 무엇이 있을까요? 그쪽과 관련해서 코드로 구현한 이유를 설명해주면 좋을거같아요!

KyoPak replied:
코드로 구현을 해보고 싶었던 이유는 일단 코드로 구현을 한다면 협업할때 충돌이 적게일어난다해서 체험을 해보고 싶었습니다..코드로 오토레이아웃을 표현해보니 조금더 명확하다고 느껴졌고 어떤 제약을 줬는지 파악하기가 매우 수월하였습니다.
하지만, 코드로 했을때 현재까지 어떻게 오토레이아웃이 그려지고 있는지 파악하기가 힘들었고 에러를 찾는 과정도 쉽지 않았던 것 같습니다. 때문에 향후에 코드로 오토레이아웃을 구성해야할 경우 조금의 밑그림이라도 생각을 하고 진행해야겠다?는 생각이 들었습니다 🤣

lina0322 commented on Expo1900/Expo1900/Common/AssetName.swift:
```diff
@@ -0,0 +1,15 @@
+    static let expoImage = "poster"
+    static let flagButtonImage = "flag"
+}
+
+
```
의미없는 줄바꿈은 지워주시면 좋겠네요!
enum으로 nameSpace를 지정하는 것과 struct로 하는 것은 어떤 차이가 있나요?
왜 enum을 선택하셨는지 궁금합니다🫡

KyoPak replied:
수정완료했습니다!
struct로 nameSpace를 지정하게되면 `private init()` 처리를 해서 불필요한 인스턴스 생성을 막아줘야 하는데
enum을 사용한다면 이러한 번거로운 점이 없어져서 사용하였습니다.

tottalE replied:
[173c6f9](https://github.com/yagom-academy/ios-exposition-universelle/pull/217/commits/173c6f92b1a741b8fceac76f16251da91fd07039)

여기에 수정했습니다!

lina0322 commented on Expo1900/Expo1900/Common/Constant.swift:
```diff
@@ -0,0 +1,11 @@
+//  Created by Kyo, tottalE on 2022/10/19.
+//
+
+enum Constant {
+    static let nextButtonTitle = "한국의 출품작 보러가기"
+    static let koreaItemsNavigationTitle = "한국의 출품작"
+}
```
모든 화면에서 사용되는 상수인가요??

tottalE replied:
한번에 스트링 값들을 모아서 보기 위해서 네임 스페이스를 생성해 두었는데, 각 뷰 컨트롤러에 값으로 두는 것이 좋은지 엘림의 의견이 궁금합니다! 각 뷰컨트롤러에 두면 흩어질 것 같은 단점이 있는 것 같아서요!

lina0322 replied:
장단점은 어떤 경우에든 있겠지만, 예를 들어서 지금 같은 경우 nextButtonTitle 화면 하나만 더 생겨도 변수명이 애매해지지 않을까요??
필요한 것은 필요한 곳에 존재하고, 각각의 역할을 명확하게 하는게 좋은 것같습니다.
지금 같은 경우에서는 어디든지 Constant를 입력하면 접근 가능한데, 굳이 필요없는 곳에서 접근할 필요가 있나요?🫣

KyoPak replied:
그러면 크게 반복해서 사용되지않는 nameSpace 같은 경우는 해당 String이 필요한 부분에 직접적으로 넣어주거나 필요한 부분에 따로 프로퍼티 생성하는 쪽이 더 명확하다는 의미가 맞으실까요??
-  저는 Constant로 따로 관리를 하고 엘림 조언대로 변수명을 바꾸는 방향으로 생각을해서 여쭤봅니다!🥲

lina0322 replied:
해당 스트링이 필요한 부분에 직접 넣어주는건 아닌것 같고
그 뷰컨이 해당 enum을 들고있으면 될거같습니다!

tottalE replied:
네 아래와 같이 분리해 주었습니다!

[852db12](https://github.com/yagom-academy/ios-exposition-universelle/pull/217/commits/852db123e693dd74e9e74c053deee77574d48d13)

lina0322 commented on Expo1900/Expo1900/Common/DataError.swift:
```diff
@@ -0,0 +1,26 @@
+//  Created by Kyo, tottalE on 2022/10/19.
+//
+
+enum DataError: Error {
+    case noneDataError
+    case decodingError
+    case unknownError
+    case formattingError
+    
+    var message: String {
+        switch self {
+        case .noneDataError:
+            return "NoneData Error"
+        case .decodingError:
+            return "Decoding Error"
+        case .formattingError:
+            return "Formatting Error"
+        case .unknownError:
+            return "Unknown Error"
+        }
+    }
+}
```
error와 관련된 프로토콜 중에 description 등을 리턴하는 것이 있는데 한 번 찾아보면 도움이 될거같네요!👀

tottalE replied:
네 찾아본 결과 아래와 같이 수정해 주었습니다!
[565b2bc](https://github.com/yagom-academy/ios-exposition-universelle/pull/217/commits/565b2bc52a3605219dc0263cc6b73eb0ff487a55)

저희가 드는 의문점이 있었는데요.
1. NSLocalizedString으로 꼭 리턴을 해주어야 하는지 의문이 들었습니다. 그래서 우선 그냥 스트링으로 리턴을 해주었는데 잘 동작하는 것을 확인해 줄 수 있었습니다.
2. 정확히 localizedError가 어떻게 동작하는지 이해하기 어려웠습니다. errorDescription을 바꾸어 주면 localizedDescription이 바뀌는 이유가 정확히 이해되지는 않았습니다. 조언 부탁드려도 될까요?

lina0322 replied:
1. String으로 반환하는 것이 문제되지는 않습니다! 하지만 일정 프로토콜을 만족시킨다면, 다른 부분에 적용시키기가 더 쉽고 일반적으로 사용할 수 있어서 조언드렸습니당!
2. 요거는 질문이 이해가 안가는데요, [블로그 글](https://ios-development.tistory.com/849)과 관련이 있을까요??

tottalE replied:
failureReason 을 적용하여 해결해 주었습니다! 감사합니다!

lina0322 commented on Expo1900/Expo1900/Common/Identifier.swift:
```diff
@@ -0,0 +1,11 @@
+//  Created by parkhyo on 2022/10/19.
+//
+
+enum Identifier {
+    static let koreaItemsVCIdentifier = "koreaItemsVC"
+    static let cellIdentifier = "itemCell"
+}
```
앱 전체에서 들고 있을 필요가 있나요? 
cell은 cell이 알아서 들고있고, viewController는 viewController가 들고있을 수 있는 방법은 없을까요?

KyoPak replied:
엘림 조언대로 Identifier는 앱 전체에서 들고 있을 필요가 없을 것 같아 각 필요한 부분에 상수 프로퍼티로 지정해서 사용했습니다!
[0214a96](https://github.com/yagom-academy/ios-exposition-universelle/pull/217/commits/0214a96f4d78b91bd28246bf08baccabb258bb5e)

lina0322 replied:
상수 프로퍼티 좋습니다! 가능하면 프로토콜로 만드는게 좋을것 같은데 어렵다면 이번엔 패스해도 괜찮아요

lina0322 replied:
아, 지금 확인했는데 필요한 부분에서 상수로 들고있는게 아니라
그 이름을 가진 셀, 혹은 뷰컨 자체가 들고있는 방향으로 해볼 수 있을까요?🙏

KyoPak replied:
```swift
final class KoreaItemsViewController: UIViewController {
    private let cellIdentifier = "itemCell"
```
위와 같은 형태로 구현하였는데 혹시 엘림의 의도와 벗어날까요??

lina0322 replied:
네, UITableViewCell 내부에 cellIdentifier가 있는 구조입니다! 
이유는 화면이 비슷한 경우에 하나의 셀을 만들어서 여러 화면에서 사용하게 되는데,
그 때마다 저렇게 만들어 주는 것 보다 셀 자체가 들고 있는게 더 좋고
셀 네이밍을 변경하거나 할때 자동적으로 모든 화면에 적용되기 위함입니다!

KyoPak replied:
```swift 
func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cellIdentifier = "itemCell"
        guard let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier) else {
```
엘림 말을 듣고보니 내부 전체에서 접근하게끔 하기 보다는 위와 같이 메서드 내부에서 사용하게끔 하는게 더 좋겠네요! 
제가 이해한 내용이 맞을까요??

lina0322 replied:
아닙니다.. UITableViewCell가 들고 있는겁니다.
지금 말씀대로라면 마찬가지로 각 화면마다 컨트롤해야하지 않나용?

```swift
class SomeTablaViewCell: UITableViewCell {
       // 여기에 선언
}
```

음, 지금 이해가 안가는게 어떤 부분일까요?? 왜 이렇게 하는지를 잘 모르겠나요, 아니면 구현 방식을 모르겠나용??

KyoPak replied:
아..제가 UITableViewCell을 직접 만들지 않고 ViewController내부에서 기본Cell을 사용해서 헷갈렷던것같습니다! 
그러면 지금 같은 경우에는 UITableViewCell을 따로 만들고 그 내부에서 Identifier 및 cell관련된 오토레이아웃 등을 관리 해주는게 맞겠네요! 혹시 맞을까요..?

lina0322 replied:
넵, 이 방향이 좋을 것 같습니다!!ㅎㅎ 제가 설명이 좀 부족했네요🥹
사실 이 방향만 잡아드리려고 했고, 아까처럼 메서드 내부에서만 하는것도 지금 구조상으로는 문제가 없어서 수정할 필요는 없는데
이게 앱이 커지고, 같은 셀을 여기저기서 사용하게 될때, identifier나 cell 이름이 변경되면 직접 다 찾아가서 변경도 해야하고 
잘못되어있는 경우에 빌드가 안되거나(이러면 다행인데), 앱이 터지는 경우가 있어서욥!
아직 초반이라 이렇게까지 안나가도 되지만, 코드로 작업하시는 것도 있고, 미리 알아두셔도 좋을 것 같아서 말씀드렸습니당~

KyoPak replied:
엘림 조언대로 CustomCell을 구현해서 내부에 Identifier nameSpace를 넣어주었습니다. 타입 프로퍼티를 사용하는 방법을 잊고있었네요..!👍

lina0322 commented on Expo1900/Expo1900/Controllers/DetailViewController.swift:
```diff
@@ -0,0 +1,120 @@
+
+import UIKit
+
+final class DetailViewController: UIViewController {
```
final 키워드를 사용하셨네요!
어떤 장점이있나요?🫣

KyoPak replied:
final 키워드를 사용하면 재정의 및 상속이 불가해집니다! 또한 , 클래스의 Dynamic Dispatch 동작이 Static Dispatch로 동작해서 성능이 향상되는 것으로 알고 있습니다.

lina0322 replied:
오, Dynamic Dispatch와 Static Dispatch의 차이는 무엇인가요?👀

KyoPak replied:
제가 알고 있기로는 Static Dispatch의 경우에는 컴파일 시점에 실행할 함수의 메모리 주소를 코드자체에 삽입해서 빠른 것으로 알고 있습니다.
하지만 Dynamic Dispatch는 V-table에 함수의 포인터를 배열형태로 보관하기 때문에 속도가 Static Dispatch보다는 느린 것으로 알고 있습니다.
하지만 v-table에 각 함수들의 포인터들을 저장해놓을 수있기 때문에 상속, 재정의가 가능한 것으로 알고있습니다.!

lina0322 commented on Expo1900/Expo1900/Controllers/DetailViewController.swift:
```diff
@@ -0,0 +1,120 @@
+
+final class DetailViewController: UIViewController {
+    
+    var koreaItem: KoreaItem?
```
혹시 밖에서 접근 할 일이 있나용?? 외부에서 변경되면 안 될 것 같은 데이터인데요!

tottalE replied:
```swift
extension KoreaItemsViewController: UITableViewDelegate {
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let detailVC = DetailViewController()
        detailVC.koreaItem = koreaItems[indexPath.row]
        navigationController?.pushViewController(detailVC, animated: true)
    }
}
```
위와 같이 KoreaItemsViewController 뷰 컨트롤러 내에서 접근하여 값을 주입해 주기 때문에 internal으로 남겨 두었습니다! 혹시 문제되는 부분이 있을까요?

lina0322 replied:
아하 밖에서 주입받는 데이터군요.🤗
혹시 init할때 받는건 어떻게 생각하세요?

tottalE replied:
네 저희도 init으로 받는 방법을 탐구해보다가 실은 마땅한 해결책을 찾지 못한 상태였어서 일단은 저렇게 두었습니다!

KyoPak replied:
엘림! 해당 부분 init 사용해서 private처리 완료하였습니다!
지난 번에 시도했을때 안된이유가 UIViewController의 지정생성자를  호출했어야 했는데 편의생성자인 `super.init()`을 호출해서 안되는 거였네요!
`super.init(nibName: nil, bundle: nil)` 해당 지정생성자 호출해서 해결하였습니다!
[d185d2a](https://github.com/yagom-academy/ios-exposition-universelle/pull/217/commits/d185d2aff6dbc1846b201db39d2aac4c88bfff1c)

lina0322 replied:
좋습니다. 멋집니다! 왜 안됐는지도 잘 기억해두시면 좋겠어요🫡
(기존 방식도 잘못된건 아닙니다!! 기존 방식도 자주 사용해요!)

lina0322 replied:
여러분, 근데 요거 좀 오해 소지가 있을거같ㅇㅏ서.. [이 글](https://velog.io/@dev_jane/UIViewController-%EC%84%9C%EB%B8%8C%ED%81%B4%EB%9E%98%EC%8A%A4%EC%9D%98-custom-initializer-%EB%A7%8C%EB%93%A4%EA%B8%B0required-initializer-initcoder-must-be-provided-by-subclass-of-UIViewController) 한 번 읽어보시면 좋을거같아요.
제가 글로 설명하려했는데, 너무 잘써주신 분이 계셔서ㅎㅎㅎ🎉

KyoPak replied:
Step-3 진행하면서 주신 글 읽어보겠습니다! 👍

lina0322 commented on Expo1900/Expo1900/Controllers/DetailViewController.swift:
```diff
@@ -0,0 +1,120 @@
+    private let itemImageView: UIImageView = {
+        let itemImageView = UIImageView()
+        itemImageView.contentMode = .scaleAspectFit
+        itemImageView.translatesAutoresizingMaskIntoConstraints = false
```
translatesAutoresizingMaskIntoConstraints는 어떤 역할을 하나요?ㅎㅎ

tottalE replied:
오토 리사이징을 키고 끄는 기능을 수행합니다. 오토 리사이징의 경우 Autoresizing mask를 사용하여, superview가 커지거나 줄어듦에 따라 subview의 크기나 위치를 조정하는 것을 이야기 하는데 완전히 이해되어 있는 상태는 아닌 것 같습니다! 스토리보드의 경우 기본값이 false이지만 코드상에서는 true이기 때문에 조정을 해 주었습니다!

lina0322 commented on Expo1900/Expo1900/Controllers/DetailViewController.swift:
```diff
@@ -0,0 +1,120 @@
+    }()
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
```
super.viewDidLoad()를 호출하지 않으면 어떻게 될까요?

KyoPak replied:
실험해본 결과, 이상없이 잘 작동되었습니다!
하면서 의문점이 생겼습니다. 그 동안 super.viewDidLoad()를 항상 호출해왔는데 내부 정의를 보니 비어있는 함수 였습니다.
항상 super로 호출을 해주는 것을 권장하는 이유가 혹시 기능적으로는 이상이 없지만
OOP법칙 리스코프치환법칙 위배하지 않도록 super.viewDidLoad(), super.viewWillAppear()등 을 호출하는 것인가 해서요..!
엘림의 의견이 궁금합니다 :)

lina0322 replied:
요거 어디까지 대답해드려도 될지 모르겠는데,
제 기억에 수업때 나왔던 내용이라 확인차 여쭤봤는데, 앱 라이프 사이클 관련해서 수업 진행 되었나용?!

KyoPak replied:
배웠는데 깜빡했었습니다..! 리스코프 치환법칙 지킬겸.. 기능도 호출할 겸 .. super.viewDidLoad()를 호출하는 것 같습니다!

lina0322 commented on Expo1900/Expo1900/Controllers/DetailViewController.swift:
```diff
@@ -0,0 +1,120 @@
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        setUI()
+        setupConstraint()
```
set? setup? 어떤 의미 차이로 사용하고 있나요?👀

tottalE replied:
저희는 일단 setup은 constraint를 고정해줄때 화면을 셋업해준다는 느낌으로 constraint에만 사용해 주었습니다. set의 경우 UI에 데이터를 넣거나 배치하는 등에 사용을 해주었습니다. 통상적인 기준이 있을까요?

lina0322 replied:
보통 준비하는 과정에서는 setup을 더 많이 사용하는 것 같습니다.
사바사 회바회겠지만, configure을 사용하는 경우도 많고
명확한 기준이 기억이 안나는데, get/set은 예약어도 있고해서 잘 사용하지 않는것으로 알고있습니다~

tottalE replied:
네 다음부터는 setup 위주로 사용해 주겠습니다! 코멘트 감사합니다!

lina0322 commented on Expo1900/Expo1900/Controllers/DetailViewController.swift:
```diff
@@ -0,0 +1,120 @@
+    }
+}
+
+private extension DetailViewController {
+    
+    private func setUI(){
```
extension에 private을 하면 그 블럭의 내용들의 접근제어자는 어떻게 설정되나요??

KyoPak replied:
- 일단 extension에 접근제어를 붙히게 될 경우, 일단 본문의 접근제어를 따라가게됩니다. 그리고 본문의 접근제어보다 낮은 수준일 경우는 자동으로 본문의 접근제어를 따라가는 것으로 확인하였습니다.
- 그리고 extension에 private를 붙힐 경우, 내부 블럭의 내용들은 접근제어들은 `fileprivate`로 되는 것으로 확인되었습니다! 
혹시 제가 확인한 내용이 맞을까요👀

lina0322 replied:
제 질문의 의도는, extension에도 붙이고 내부 함수에도 붙여서 이중으로 되어있기에 궁금했습니다🧐

KyoPak replied:
아 사실 테스트를 해보면서 함수 내부에는 굳이 필요가 없을 것 같다고 생각은 했었습니다..! extension을 달면 내부적으로도 fileprivate처리가 되긴하더라구요! 해당 부분 토털이와 의논후 수정해보겠습니다 :)

KyoPak replied:
의논 결과, extension으로 내부 func들을 전부 접근제어 기능을 사용하게 할 수 있지만 아직까지는  func 하나하나 접근제어를 명시해주는게 좋을 것 같아 extension앞에 private를 제거하였습니다!

lina0322 commented on Expo1900/Expo1900/Controllers/DetailViewController.swift:
```diff
@@ -0,0 +1,120 @@
+    }
+    
+    private func setNavigationBar() {
+        self.title = koreaItem?.name
```
self를 붙이는 기준이 궁금합니다~

tottalE replied:
자신의 인스턴스의 프로퍼티를 지칭할때 self를 붙여줍니다!

lina0322 commented on Expo1900/Expo1900/Controllers/DetailViewController.swift:
```diff
@@ -0,0 +1,120 @@
+        super.viewDidLoad()
+        setUI()
+        setupConstraint()
+    }
+}
+
+private extension DetailViewController {
+    
+    private func setUI(){
+        setNavigationBar()
+        setData()
+        setViews()
+    }
+    
+    private func setupConstraint(){
+        setupScrollView()
+        setupStackView()
+        setupContentViewElement()
+    }
+    
+    private func setNavigationBar() {
+        self.title = koreaItem?.name
+        let appearance = UINavigationBarAppearance()
+        navigationController?.navigationBar.scrollEdgeAppearance = appearance
```
appearance는 어떤 이유로 사용하신걸까용?🙂

KyoPak replied:
- `UINavigationBarAppearance`를 사용함으로서 Bar의 모습을 지정해줄 수 있는 것으로 알고 있습니다. 해당부분을 사용 안할 경우 Bar의 Line이 생성되지 않아서 사용하였습니다.


- 위와 같이 했을 경우 테스트로 BackGroundColor을 지정해 줬을 때는 스크롤 할 경우는 Bar의  BackGroundColor가 나오지 않아 `standardAppearance`라는 것을 알게 되었는데 Line은 `standardAppearance`을 지정해주지 않아도 스크롤 할 경우 생성이 되는데 그것에 대한 이유는 아직 잘 모르겠습니다..😂

KyoPak replied:
`standardAppearance`을 지정해주지 않아도 스크롤 할 경우 Line이 생성이 되는데 그것에 대한 이유를 찾지 못해서 일단 
standardAppearance, scrollEdgeAppearance 동일하게 하기 위해 아래 코드 추가하였습니다!
```swift
navigationController?.navigationBar.standardAppearance = appearance
```

lina0322 commented on Expo1900/Expo1900/Controllers/DetailViewController.swift:
```diff
@@ -0,0 +1,120 @@
+    }
+    
+    private func setViews() {
+        view.backgroundColor = .white
```
backgroundColor를 설정하지 않으면 어떤 색상이 나오나요??

tottalE replied:
한번 실험해 보니, 검정 background가 나오는 것을 확인했습니다!

lina0322 commented on Expo1900/Expo1900/Controllers/DetailViewController.swift:
```diff
@@ -0,0 +1,120 @@
+            textView.leadingAnchor.constraint(equalTo: stackView.leadingAnchor, constant: 10),
+            textView.trailingAnchor.constraint(equalTo: stackView.trailingAnchor, constant: -10),
+            textView.bottomAnchor.constraint(equalTo: stackView.bottomAnchor)
+        ])
```
마이너스로 constant를 지정하는 부분은 이해가 잘 되셨나요?ㅎㅎ
가끔 코드로 짜면 헷갈리는 부분 중에 하나인데요.
이 constant 값도 따로 관리해주시면 좋겠네요!

tottalE replied:
네 constant 값 따로 분리해 주었습니다 ㅎㅎ

[4086a8e](https://github.com/yagom-academy/ios-exposition-universelle/pull/217/commits/4086a8e68edbd2fdccd6f3783f48bb970a2bcbd3)

`textView.trailing = stackView.trailing - 10`으로 등식이 맞아지기 때문에 등식을 기준으로 생각했을때 이해가 편했습니다. 여러모로 코드로 constraint를 잡아주는게 이해가 편한 이유이기도 한 것 같아요 ㅎㅎ

lina0322 commented on Expo1900/Expo1900/Controllers/ExpoViewController.swift:
```diff
@@ -0,0 +1,63 @@
+    @IBOutlet private weak var visitorLabel: UILabel!
+    @IBOutlet private weak var placeLabel: UILabel!
+    @IBOutlet private weak var timeLabel: UILabel!
+    @IBOutlet private weak var descriptionTextView: UITextView!
+    @IBOutlet private var buttonImage: [UIImageView]!
+    @IBOutlet private weak var nextButton: UIButton!
```
weak를 붙이는 것과 안붙이는건 어떤 차이가 있나요?ㅎㅎ

KyoPak replied:
일단 weak키워드는 참조하고 있는 객체의 RC를 올리지 않는 것으로 알고 있습니다.
- weak을 붙히지 않는다면 ViewController에서 각 view객체들을 참조하고 있으므로 RC가 1씩 상승하게 됩니다. 그리고 ViewController의 view가 view객체들을 또 참조하고 있으므로 RC가 총 2가 되게 됩니다! (실제로 테스트결과 여기저기서 RC를 올리는 것 같았습니다..RC가 막 6, 7, 8 찍히더라구요)
- weak을 붙히게 된다면 ViewController가 각 view객체들을 참조해도 RC를 올리지 않습니다.
- 따라서 추후에 메모리가 부족해져서 view를 내릴 상황에 view와 view객체들도 같이 메모리에서 해제될 수 있게 됩니다!

lina0322 commented on Expo1900/Expo1900/Controllers/ExpoViewController.swift:
```diff
@@ -0,0 +1,63 @@
+        setupElementWithData()
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
+        navigationController?.setNavigationBarHidden(true, animated: true)
+    }
+    
```
네비게이션을 감춰주고 보여주는건 한 세트로 같이 있는게 나을 것 같습니다.
차후에 기획이 바뀌거나 기능이 추가되어 이 페이지나, 다른 페이지가 새로운 경로에서 호출될 수도 있으니까요!

tottalE replied:
감춰주고 보여주는 것이 한 세트로 있어야 한다는 것이 어떤 의미인지 이해가 잘 가지 않는데 부연설명 부탁드려도 될까요?
현재 저희 코드에서 첫번째 화면에서는 네비게이션을 감춰주고 두번째 화면에서는 보여주고 있는데 이걸 한세트로 합치면 생각한 UI와 다른 결과가 나올 것 같아서요!

lina0322 replied:
두번째 화면에서 들어갈때 보여주고, 나올때 제거할 수도 있지 않나용?
네비게이션이 전체적으로 다 적용되는거면 필요하지 않은 화면에서 보여주고 없애고 하는게 맞고
전체적으로 다 보여지지 않는다면, 보여지는 화면 자체에서 보여주고 없애고 하는게 맞지 않을까요??

lina0322 commented on Expo1900/Expo1900/Controllers/ExpoViewController.swift:
```diff
@@ -0,0 +1,63 @@
+                print(error.message)
+            } else {
+                print(DataError.unknownError.message)
+            }
```
사용자에게 배포되는 서비스에 print 함수가 들어갈 필요가 있을까요?
디버깅 시에만 콘솔에 찍히게 하려면 어떻게 해야 좋을까요.
또 사용자는 에러가 나면 화면이 그냥 아무것도 안 나와서 당황하진 않을까요??👀

tottalE replied:
- 디버깅 시에만 콘솔에 찍히도록 수정하였고,
[394b600](https://github.com/yagom-academy/ios-exposition-universelle/pull/217/commits/394b600e1976df22ae30a74caf8a347d3e22d632)

- 에러 발생시 alert이 뜨도록 수정해 주었습니다!
[a775808](https://github.com/yagom-academy/ios-exposition-universelle/pull/217/commits/a775808e42e12d9701a392269ab2fdb51aa5650e)

KyoPak replied:
해당 Alert 사진 첨부드립니다.
![스크린샷 2022-10-25 오후 12 59 04](https://user-images.githubusercontent.com/59204352/197679438-72d3fb87-5384-4b42-95e3-ee06c93ce2a8.png)

추가로 Alert의 message 부분에 LocalizedError의 프로퍼티인 failReason을 사용하였습니다. 처음에는 NSLocalizedString의 comment를 사용하려했는데 잘 되지않아 failReason을 사용하였습니다.

lina0322 replied:
`NSLocalizedString의 comment를 사용하려했는데 잘 되지않아`라고 말씀하셨는데, 왜일까용? 어떻게 안됐는지 궁금해요.

KyoPak replied:
공식문서 찾아보니깐 `comment` 파라메터는 나중에 strings 파일화를 할 때 Key-Value위에 주석을 위한 파라메터라고 합니다. 
때문에 따로 빼서 사용하는 기능이 아니라서 그런지 예시나 메서드들을 찾을 수가 없었습니다..
```
comment
The comment to place above the key-value pair in the strings file. 
This parameter provides the translator with some context 
about the localized string’s presentation to the user.
```

lina0322 commented on Expo1900/Expo1900/Controllers/KoreaItemsViewController.swift:
```diff
@@ -0,0 +1,87 @@
+extension KoreaItemsViewController: UITableViewDelegate {
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let detailVC = DetailViewController()
+        detailVC.koreaItem = koreaItems[indexPath.row]
```
VC말고 풀어서 적는 습관을 길러주세요.

논리적으로는 그럴일이 없겠지만, 만약에 koreaItems[indexPath.row]에 들어가는 indexPath.row숫자가 잘못되어
out of range 에러가 발생하면 어떻게하죠? 그런 일을 방지할 수 있는 방법은 무엇이 있을까요?

tottalE replied:
VC 는 아래와 같이 수정해 주었습니다.
[f174641](https://github.com/yagom-academy/ios-exposition-universelle/pull/217/commits/f174641136ba4d161e461183802a55d59aaef511)

KyoPak replied:
```swift
if indexPath.row < koreaItems.count {
    let detailViewController = DetailViewController(koreaItem: koreaItems[indexPath.row])
    ....
}
```
실제 코드에는 반영하지 않았지만 out of range 에러를 방지하고자 하면 위와 같이 조건문을 추가하면 되지않을까 싶습니다.
혹시 현업에서는 실제로 TableView 관련해서 out of range 에러등을 고려하나요??👀

lina0322 replied:
배열은 언제든지 터질 수 있기 때문에, 테이블뷰라는 특정한 상황이 아니어도 배열에 인덱스로 접근하는건 항상 신경써야 한다고 생각합니다.
배열에 안전하게 접근하는 방법에 대해서 좀 더 찾아보셔도 좋을거같습니다!

KyoPak replied:
해당 내용은 Step-3 진행 후 한번 고려해보겠습니다 ! :)

lina0322 commented on Expo1900/Expo1900/Controllers/KoreaItemsViewController.swift:
```diff
@@ -0,0 +1,87 @@
+            return errorCell
+        }
+        
+        var content = cell.defaultContentConfiguration()
```
오랜만에 보네요ㅎㅎㅎ
아무래도 어느 곳이든.. 기본을 사용하는 경우는 거의 없어서 현업에서는 defaultContentConfiguration를 사용하실 일은 없을거에요.

근데 위쪽에 error cell은 왜 defaultContentConfiguration로 안하셨나요?

tottalE replied:
에러 셀은 간단히 에러인 것만 보여주기 위해서 config까지는 적용해 주지 않았습니다!
이 부분도 config를 해주는 것이 좋을까요? ㅎㅎ

lina0322 replied:
ㅎㅎ... 저는 defaultContentConfiguration 자체를 잘 사용하지 않아서, 통일성을 생각해서 여쭤봤습니당
상관없습니다!

lina0322 commented on Expo1900/Expo1900/Controllers/KoreaItemsViewController.swift:
```diff
@@ -0,0 +1,87 @@
+        content.secondaryText = koreaItems[indexPath.row].shortDescription
+        content.secondaryTextProperties.font = UIFont.systemFont(ofSize: 15)
+        
+        content.image = UIImage(named: koreaItems[indexPath.row].imageName)
+        content.imageProperties.maximumSize = CGSize(width: 80, height: 80)
+        content.imageProperties.reservedLayoutSize = CGSize(width: 80, height: 80)
+        content.imageToTextPadding = CGFloat(10)
```
image가 가운데 있도록 위치하고 싶은건가요?
imageView에 이미지를 정렬하는 속성값들에 대해서는 알고 계실까요??

tottalE replied:
곧 depreciated될 cell으로 직접 imageView에 접근하는 방식에서는
```swift
        cell.imageView?.centerXAnchor
```
위와 같이 접근해서 가운데 위치해줄 수 있을 것이라고 생각이 드는데, defaultContentConfiguration을 써준 상태에서는 속성값이 없어서 위와 같이 작성해 주었습니다!

lina0322 commented on Expo1900/Expo1900/Models/DataManager.swift:
```diff
@@ -0,0 +1,35 @@
+
+import UIKit
+
+struct DataManager<T: Decodable> {
```
DataManager는 상당히 광대해보여요.
실제로 viewModel처럼 DataManager를 사용하는 일이 많아서 네이밍이 적절한것 같진 않고요
DecodeManger라던가 다른 이름을 한번 생각해보시면 좋겠네요!

tottalE replied:
DecodeManger 이름 너무 좋은 것 같습니다! 말씀하신거 반영해서 네이밍 수정했습니다 ㅎㅎ

[cca8c96](https://github.com/yagom-academy/ios-exposition-universelle/pull/217/commits/cca8c96279e3e8faa5d691c262b520b8d1c991bb)

lina0322 commented on Expo1900/Expo1900/Models/DataManager.swift:
```diff
@@ -0,0 +1,35 @@
+    
+    private let decoder = JSONDecoder()
+    
+    func fetchData(name: String) throws -> T? {
```
오 제네릭을 사용하셨네요. 제네릭은 무엇인가요??

KyoPak replied:
여러 타입 별로 메서드나 클래스 등을 만들어줄 필요없이 여러타입에서 호환이 되게 끔 하는 구현하는 방법을 제네릭이라고 합니다!
해당 부분은 `ExpositionIntroduction`, `KoreaItem` 타입이 호환 가능하게 끔 제네릭을 사용하였습니다!

lina0322 commented on Expo1900/Expo1900/Models/DataManager.swift:
```diff
@@ -0,0 +1,35 @@
+        return introData
+    }
+    
+    func fetchDataList(name: String) throws -> [T] {
```
T로만 해서 위 함수만 사용할 순 없나요?

KyoPak replied:
해당 내용을 생각해보고 실제로 fetchDataList 함수 하나로 합치려고 시도를 했었습니다.
하지만 Data의 틀이 expo JSON에서는 단일 데이터 형태이고,  koreItems JSON에서는 리스트 형태의 데이터였기 때문에 구분하는게 더 맞다고 판단이 들었습니다!

lina0322 commented on Expo1900/Expo1900/Models/ExpositionIntroduction.swift:
```diff
@@ -4,11 +4,44 @@
 //
 //  Created by tottalE, Kyo on 10/17/22.
 //
+import Foundation
```
Foundation을 다시 넣으셨네요.
어떤 기능을 가지고있나요?

KyoPak replied:
Formatter 기능을 사용해서 Foundation을 import했습니다.
Foundation기능은 기본적인 데이터 타입(String, Int) 부터 자료구조, 구조체나 타이머, 네트워크 통신, 파일 관리 등 기본적이나 프로그램의 중심을 담당하는 것으로 알고 있습니다.

lina0322 commented on Expo1900/Expo1900/Models/ExpositionIntroduction.swift:
```diff
@@ -4,11 +4,44 @@
+        case visitorNumber = "visitors"
+    }
+    
+    private func formatVisitorNumber() throws -> String {
+        let formatter: NumberFormatter = NumberFormatter()
+        formatter.numberStyle = .decimal
+        guard let formattedVistorNumber = formatter.string(for: visitorNumber) else { throw DataError.formattingError }
+        return formattedVistorNumber
+    }
```
NumberFormatter나 Int를 extension하는건 어떨까요?

tottalE replied:
네 Formatter와 Int를 extension 해준 후에 computed property로 접근할 수 있도록 해주었습니다! 코드가 깔끔해졌네요!

[2566545](https://github.com/yagom-academy/ios-exposition-universelle/pull/217/commits/256654530c45ccf266cc7acfd8e22b694daf833e)

lina0322 commented on Expo1900/Expo1900/Models/ExpositionIntroduction.swift:
```diff
@@ -4,11 +4,44 @@
+                print(DataError.unknownError.message)
+            }
+        }
+        return "방문객 : \(formattedVistorNumber) 명"
+    }
+    var locationDescription: String {
+        return "개최지 : \(self.location)"
+    }
+    var durationDescription: String {
+        return "개최 기간 : \(self.duration)"
+    }
```
지금도 괜찮지만 string format을 한 번 공부해볼까요?

tottalE replied:
네 스트링 포매터 적용해서 수정해 보았습니다 ㅎㅎ

[36b64a9](https://github.com/yagom-academy/ios-exposition-universelle/pull/217/commits/36b64a9632f3f8970db0398b8774b894181ec51c)

lina0322 commented on Expo1900/Expo1900/Models/KoreaItem.swift:
```diff
@@ -5,13 +5,13 @@
     let description: String
     
-    enum CondingKeys: String, CodingKey {
+    private enum CodingKeys: String, CodingKey {
```
LGTM👏

KyoPak committed:
docx: ReadME 실행영상 50% 조절

lina0322 reviewed (changes requested):
두 분 모두 고생하셨어요!
전체적으로 코드가 깔끔하게 정돈되어있어서 읽기 좋았습니다.
scrollView와 관련된 부분은 추가로 찾아본 결과가 아주 좋네요.
스토리보드를 통해서도 scrollView를 만들어보시고 한 번 시도해보세요~
눈으로 보면서 확인하는 것도 이해에 도움이 될거같네요.

고민하신 부분은 step3에서 볼 수 있길 기대해볼게요.
또 컨벤션은 워낙 회사에 따라 달라지기 때문에 지금처럼 가셔도 무난할거같고요
extension은 프로토콜 별로 나누시는 분도 계시고, 기능 별로 나누거나 접근제한자 별로 나누거나 다양한 방법이 있어요.
다양하게 시도해보시고 다른 분들 코드 보면서 좋은거 따라해보시면 좋을 것 같네요
답이 있는 부분이 아니니 다양하게 경험해보시길 바랍니다~

달아드린 코멘트 중에 이해 안가거나 추가로 궁금한거 있으시면 언제든지 디엠주세요!

tottalE committed:
refactor: AssetName 줄 바꿈 수정

KyoPak committed:
refactor : Identifier 파일 삭제 및 각 Identifier 필요 부분에 직접 대입

KyoPak committed:
refactor : DetailViewController 내부 koreaItem 프로퍼티 private 접근제어 처리

- DetailViewController 내부 koreaItem 프로퍼티 private 접근제어 처리
- 해당 부분 private 처리에 따른 Data 전달을 위한 init 추가

tottalE committed:
refactor: ExpositionIntroduction 모델의 description 값들에  String format 구현

KyoPak committed:
refactor : Identifier 프로퍼티 private 접근제어 처리

tottalE committed:
refactor: DecodeManger로 DataManger 변경

KyoPak committed:
refactor : DataError extension하여 LocalizedError 프로토콜 채택

- DataError extension하여 LocalizedError 프로토콜 채택
- 기존 message 연산 프로퍼티 대신 프로토콜 채택하여 errorDescription 연산프로퍼티로 구현

tottalE committed:
refactor: DetailViewController.swift 파일 내 Constant 네임스페이스 추가, constraint 관리

KyoPak committed:
fix : textView Edit 가능 여부 false로 변경

- textView Edit 가능 여부 false로 변경

tottalE committed:
refactor: VC에 NSLog 적용

KyoPak committed:
feat : Error 발생시 Alert 띄우기 구현 및 설명을 위한 failureReason 프로퍼티 구현

tottalE committed:
refactor: VC ViewController로 풀어서 rename

tottalE committed:
refactor: Int, Formatter extension 추가 및 ExpositionIntroduction model에 formattedWithSeparator computed property 적용

KyoPak committed:
refactor : 기본 Cell 구현을 Custom Cell 구현으로 변경

- 기본 Cell 구현을 Custom Cell 구현으로 변경
- cellIdentifier를 cell 내부에 선언

KyoPak committed:
refactor : KoreaItemTableViewCell 내부 제약조건 Constant nameSpace로 변경

lina0322 commented on Expo1900/Expo1900/Common/Extensions/Int+Extensions.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+extension Int {
+    var formattedWithSeparator: String {
+        return Formatter.withDecimalSeparator.string(for: self) ?? ""
+    }
+}
```
LGTM👏

lina0322 commented on Expo1900/Expo1900/Controllers/ExpoViewController.swift:
```diff
@@ -0,0 +1,79 @@
+    }
+    
+    @IBAction private func nextButtonTapped(_ sender: UIButton) {
+        guard let koreaItemsVC = storyboard?.instantiateViewController(withIdentifier: koreaItemsVCIdentifier) as? KoreaItemsViewController else { return }
```
여기도 VC로 사용중이네용!

KyoPak replied:
해당 VC 풀어서 적는 습관을 가져야겠네요!! 수정했습니다 👍

lina0322 commented on Expo1900/Expo1900/Controllers/ExpoViewController.swift:
```diff
@@ -0,0 +1,79 @@
+        } catch {
+            if let error = error as? DataError {
+                popErrorAlert(error: error)
+                #if DEBUG
+                NSLog(error.localizedDescription)
+                #endif
```
오, 제가 생각했던 것과는 다른 방향이지만 좋습니다! 
사실 저는 debugPrint 정도 생각했습니다ㅎㅎ

lina0322 commented on Expo1900/Expo1900/Views/KoreaItemTableViewCell.swift:
```diff
@@ -0,0 +1,90 @@
+
+import UIKit
+
+class KoreaItemTableViewCell: UITableViewCell {
```
여기도 final 붙여주세용!

KyoPak replied:
앗 깜빡했네요!! 붙혔습니다 👍

lina0322 commented on Expo1900/Expo1900/Views/KoreaItemTableViewCell.swift:
```diff
@@ -0,0 +1,90 @@
+
+class KoreaItemTableViewCell: UITableViewCell {
+    
+    static let cellIdentifier = "itemCell"
```
직접 지정하지 않고, 클래스 자신의 이름을 반환할 수 있을까요?ㅎㅎ

KyoPak replied:
```swift
static let cellIdentifier = String.init(describing: KoreaItemTableViewCell.self)
```
위와 같이 String 초기화를 사용하여 `cellIdentifier = "KoreaItemTableViewCell"`처럼 직접 값 지정이 아닌 반환하는 방식으로 구현 변경하였습니다.

lina0322 commented on Expo1900/Expo1900/Views/KoreaItemTableViewCell.swift:
```diff
@@ -0,0 +1,90 @@
+        setupConstraints()
+    }
+    
+    override func setSelected(_ selected: Bool, animated: Bool) {
+        super.setSelected(selected, animated: animated)
+    }
```
요거는 왜 override 하셨을까용?

tottalE replied:
처음에 기본적으로 생성된 코드였는데, 필요 없는 코드라고 생각되어 삭제해 주었습니다!

lina0322 reviewed (approved):
짧은 기간동안 많은 변화가 있었네요!! 정말 고생 많으셨어요~
추가로 달아드린 코멘트 관련하여 바로 수정하셔도 좋고, 원하시면 머지 후 다음 스텝에서 함께 수정 혹은 코멘트 주셔도 좋습니당~

tottalE committed:
feat: Constant 각 뷰컨트롤러로 분리

KyoPak committed:
refactor : KoreaItemTableViewCell final키워드 추가 및 naming 수정

- KoreaItemTableViewCell final키워드 추가
- ExpoViewController 내부 koreaItemsVC ➡️ koreaItemsViewController naming 수정
- Constant File 삭제

tottalE committed:
refactor: DetailViewController.swift 내부 extension 접근제어자 수정

KyoPak committed:
refactor :  DetailViewController내부 Navigation Bar 관련 standardAppearance지정 추가

tottalE committed:
refactor: KoreaItemTableViewCell.swift 내 불필요한 setSelected 삭제

KyoPak committed:
Merge remote-tracking branch 'refs/remotes/origin/step2'

KyoPak committed:
refactor : 각 ViewController, Cell Identifier 값 지정 방식에서 클래스 이름 반환하는 방식으로 변경
